Support multiple ignore paths for slather action

mgrebenets opened the pull request:
With these changes it's possible to specify multiple ignore paths as an array as well as keep backwards compatibility and use just a string.

``` ruby
    slather(
        proj: "project",
        scheme: "scheme",
        input_format: "profdata",
        ignore: ["Pods/*", "../**/*/Xcode*"],
       # ignore: "Pods/*", # also works
        simple_output: true
    )
```

This change also adds quotes when composing shell command, so it's becomes `--ignore "../**/*/Xcode*"` instead of `--ignore ../**/*/Xcode*`, which causes some troubles.

What do you think @mattdelves ?

Commit:
Support multiple ignore paths for slather action

Allow spaces in paths

ohwutup commented on lib/fastlane/actions/slather.rb:
```diff
@@ -32,7 +32,7 @@ def self.run(params)
         command += " --source-directory #{params[:source_directory]}" if params[:source_directory]
         command += " --output-directory #{params[:output_directory]}" if params[:output_directory]
-        command += " --ignore #{params[:ignore]}" if params[:ignore]
+        command += " #{params[:ignore].split('|').map { |path| "--ignore \"#{path}\"" }.join(' ')}" if params[:ignore]
```
this should be able to take in a legit array instead of having to split a string.

mgrebenets replied:
I'll have a look at that, thanks for the tip.
Was confused by `is_string: false` in other options, made me thing it's either a string or a bool.

mgrebenets replied:
Actually, I'm getting this error after simply feeding array to `ignore` parameter:

```ruby
ignore: ["Pods/*", "../**/*/Xcode*"],
```

```
/Users/maksymgrebenets/.rvm/gems/ruby-2.2.2/gems/fastlane_core-0.32.1/lib/fastlane_core/configuration/config_item.rb:56:in `valid?': 'ignore' value must be a String! Found Array instead. (RuntimeError)
````

Is `:short_option` going to help me with this somehow? Do you have any examples?

Otherwise it looks like I'd have to make changes to `fastlane_core` gem so that `config_item` would take an array as well.

mgrebenets replied:
Oh, I got it, `is_string: false` fixed the problem, now it takes in the array.

ohwutup commented:
hey @mgrebenets thanks for the PR, this is great. i just made one comment above and then it should be good to go!

mgrebenets commented:
@ohwutup I have changed it to accept an array like this: `ignore: ["Pods/*", "../**/*/Xcode*"]`.
The only concert is that this change is now a breaking change, because old code like `ignore: "Pods"` will not work any more. Probably that's why CI fails too.

ohwutup commented:
thanks for making that change @mgrebenets! 

i think you can add a check in there for the type whether its a string or array and add to the array of commands appropriately. 

let me know if this makes sense, I'm happy to make the change and push it up for you to take a look at!

mgrebenets commented:
@ohwutup Done that change with help of `kind_of?`.
Testing on my project looks good.

I'm not 100% sure about formatting and actual Ruby code, the hound is happy though.
Anyway, feel free to refactor/modify it and push the changes, I don't mind :+1:

ohwutup commented:
thanks @mgrebenets! i refactored it a bit just to avoid the nested conditional statement. https://github.com/fastlane/fastlane/compare/mgrebenets-feature/multiple-ignore-paths-for-slather?diff=unified&expand=1&name=mgrebenets-feature%2Fmultiple-ignore-paths-for-slather#diff-2e945dbb87e8c9ab109ca7795fd504f8R35

because i can't push to your fork to add to this branch, feel free to copy that code in and push it up so you can get credit for this PR :rocket: !

mgrebenets commented:
Good! Changed the PR.
Thanks for your help!

mgrebenets commented:
@ohwutup I've just rebased on top of latest `master` and fixed few failing tests. I've also used `shellescape` to allow spaces in paths. As much as I hate spaces, I know they often occur in scheme or project names; or other unexpected places. Let me know if it looks OK or needs more fixes. Thanks :+1:

ohwutup commented:
this is awesome! thanks @mgrebenets ! last thing- would you mind squashing your commits down into one? https://github.com/fastlane/countdown#submit-a-pull-request

mgrebenets commented:
Sure, @ohwutup, squashed them.

ohwutup commented on lib/fastlane/actions/slather.rb:
```diff
@@ -135,7 +139,8 @@ def self.available_options
                                        env_name: "FL_SLATHER_IGNORE",
-                                       description: "Tell slather to ignore files matching a path",
+                                       description: "Tell slather to ignore files matching a path or any path from an array of paths",
+                                       is_string: false,
```
sorry @mgrebenets, i just noticed this param, it doesn't appear to be getting used anywhere because we are inspecting the type, am i just missing where it is used?

mgrebenets replied:
I believe this is kind of type safety in `ConfigItem` module (or class?), but it only checks for `String` type.
The default is `true`, so if I take it out and pass an array to `slather` action, the error will be this

```
[08:15:51]: You passed invalid parameters to 'slather'.
[08:15:51]: Check out the error below and available options by running `fastlane action slather`
[08:15:51]: error: coverage, 'ignore' value must be a String! Found Array instead.
```

The problem is that we have a param which can be of multiple types. And even when `is_string` is `false`, we still can pass a string and it works...

Anyway, I could add a warning that will show up in fastlane log when people use `String` type. How does that sound? 

Alternatively, I could add one more parameter that takes an array of paths and add deprecation message for the old one.

ohwutup replied:
my bad! i didn't realize that was the situation! lets merge this and we can worry about the params later! :rocket:

ohwutup commented:
:shipit:
